crates/matrix-sdk/src/widget/widget_settings/mod.rs

```diff
@@ -0,0 +1,351 @@
+mod element_call;
+
+use language_tags::LanguageTag;
+use ruma::api::client::profile::get_profile;
+use url::Url;
+
+use crate::Room;
+
+mod url_params {
+    use url::Url;
+    use urlencoding::encode;
+
+    pub static USER_ID: &str = "$matrix_user_id";
+    pub static ROOM_ID: &str = "$matrix_room_id";
+    pub static WIDGET_ID: &str = "$matrix_widget_id";
+    pub static AVATAR_URL: &str = "$matrix_avatar_url";
+    pub static DISPLAY_NAME: &str = "$matrix_display_name";
+    pub static LANGUAGE: &str = "$org.matrix.msc2873.client_language";
+    pub static CLIENT_THEME: &str = "$org.matrix.msc2873.client_theme";
+    pub static CLIENT_ID: &str = "$org.matrix.msc2873.client_id";
+    pub static DEVICE_ID: &str = "$org.matrix.msc2873.matrix_device_id";
+    pub static HOMESERVER_URL: &str = "$org.matrix.msc4039.matrix_base_url";
+    pub struct QueryProperties {
+        pub(crate) widget_id: String,
+        pub(crate) avatar_url: String,
+        pub(crate) display_name: String,
+        pub(crate) user_id: String,
+        pub(crate) room_id: String,
+        pub(crate) language: String,
+        pub(crate) client_theme: String,
+        pub(crate) client_id: String,
+        pub(crate) device_id: String,
+        pub(crate) homeserver_url: String,
+    }
+    pub fn replace_properties(url: &mut Url, props: QueryProperties) {
+        *url = Url::parse(
+            &url.as_str()
+                .replace(WIDGET_ID, &encode(&props.widget_id))
+                .replace(AVATAR_URL, &encode(&props.avatar_url))
+                .replace(DEVICE_ID, &encode(&props.device_id))
+                .replace(DISPLAY_NAME, &encode(&props.display_name))
+                .replace(HOMESERVER_URL, &encode(&props.homeserver_url))
+                .replace(USER_ID, &encode(&props.user_id))
+                .replace(ROOM_ID, &encode(&props.room_id))
+                .replace(LANGUAGE, &encode(&props.language))
+                .replace(CLIENT_THEME, &encode(&props.client_theme))
+                .replace(CLIENT_ID, &encode(&props.client_id)),
```

Comment:
Yeah it's pretty unlikely to happen, same for one of the placeholders expanding to another placeholder that could cause trouble now, I just think it would be cleaner to work iteratively and look for the end of the placeholder name before finding a replacement.

Reply:
Building up the string in a for loop is definitly more performant. So I started with that.
What do you think about this usecase: `some/path/Captain_$matrix_display_names_place` where they add a plural s in the url template?